cachi2/core/package_managers/yarn_classic/project.py

```diff
@@ -0,0 +1,176 @@
+            )
+        except json.decoder.JSONDecodeError as e:
+            raise PackageRejected(
+                reason=f"Can't parse the {path.subpath_from_root} file. {e}",
+                solution=(
+                    "The package.json file must contain valid JSON. "
+                    "Refer to the parser error and fix the contents of the file."
+                ),
+            )
+
+        return cls(path, package_json_data)
+
+
+class YarnLock(_CommonConfigFile):
+    """A yarn.lock file.
+
+    This class abstracts the underlying attributes.
+    """
+
+    yarn_lockfile: lockfile.Lockfile
+
+    @property
+    def config_kind(self) -> ConfigKind:
+        """Return kind of this ConfigFile."""
+        return "yarnlock"
+
+    @classmethod
+    def from_file(cls, path: RootedPath) -> "YarnLock":
+        """Parse the content of a yarn.lock file."""
+        try:
+            yarn_lockfile = lockfile.Lockfile.from_file(path)
+        except FileNotFoundError:
+            raise PackageRejected(
+                reason="The yarn.lock file must be present for the yarn package manager",
+                solution=(
+                    "Please double-check that you have specified the correct path "
+                    "to the package directory containing this file"
+                ),
+            )
+        except ValueError as e:
+            raise PackageRejected(
+                reason=(f"Can't parse the {path.subpath_from_root} file.\n" f"{e}"),
+                solution=(
+                    "The yarn.lock file must be valid. "
+                    "Refer to the parser error and fix the contents of the file."
+                ),
+            )
+
+        if not yarn_lockfile:
+            raise PackageRejected(
+                reason="The yarn.lock file must not be empty",
+                solution=("Please verify the content of the file."),
+            )
+
+        return cls(path, yarn_lockfile.data)
+
+
+ConfigFile = Union[PackageJson, YarnLock]
+
+
+class Project(NamedTuple):
+    """A directory containing yarn sources."""
+
+    source_dir: RootedPath
+    package_json: PackageJson
+
+    @property
+    def is_pnp_install(self) -> bool:
+        """Is the project is using Plug'n'Play (PnP) workflow or not.
+
+        This is determined by
+        - `installConfig.pnp: true` in 'package.json'
+        - the existence of file(s) with glob name '*.pnp.cjs'
+        - the presence of an expanded node_modules directory
+        For more details on PnP, see: https://classic.yarnpkg.com/en/docs/pnp
+        """
+        install_config_pnp_on = self.package_json.install_config.get("pnp", False) is True
+        pnp_cjs_exists = any(
+            PurePath(file).match("*.pnp.cjs") for file in self.source_dir.path.iterdir()
+        )
+        node_modules_exists = self.source_dir.join_within_root("node_modules").path.exists()
+        is_pnp = False
+        if install_config_pnp_on or pnp_cjs_exists or node_modules_exists:
```

Comment:
```suggestion
return install_config_pnp_on or pnp_cjs_exists or node_modules_exists
```

Reply:
Derp, this is what comes from "losing track of the meta" while refactoring.

Reply:
Fixed